README

XLIBTRACE 0.2.1
===============

by Kevin Pulo, kev@pulo.com.au, and contributors
http://www.kev.pulo.com.au/xlibtrace/
Copyright (c) 2007-2009, 2020-2021
Licensed under the GNU GPL v2 or higher, as in the COPYING file.


DESCRIPTION
-----------

xlibtrace displays the interaction between X11 client programs and the
X11 client shared library (libX11.so), by showing the Xlib calls that
are made.  Output is shown in a similar style to strace(1).

xlibtrace works by using the $LD_PRELOAD dynamic linker option to
insert itself between the target X11 client program and libX11.so.
This allows it to "intercept" all calls to X functions, whereupon it
has the opportunity to print the name of the function being called,
along with any arguments and return value.

xlibtrace allows the user to investigate which X functions are being
called, when they are being called, what parameters are passed and
what the return value is.  This can be particularly useful when
analysing the behaviour of closed-source X11 programs.


DEPENDENCIES
------------

- X11R6

- The build process additionally requires the Xlib header files, as
  well as sed and awk.


INSTALLATION
------------

Standard autoconf process:  ./configure && make && make install


USAGE
-----

    xlibtrace [options] <program> [args...]

Valid options are:

  --stderr          Send output to stderr (default)
  --stdout          Send output to stdout
  -o filename       Send output to the specified file
  -a, --append      Append output to file specified by -o

  --single          Trace only the specified program, and not any
                    other subsequently forked processes.  Useful for
                    tracing xterms, etc.

  --prompt          Prompt user prior to each call.  Useful for
                    interactively stepping through the X11 calls.
  --sleep us        Sleep for the given microseconds after each call.
                    Useful for artificially slowing down the client.
  --xflush          Call XFlush() after each call.  Useful in
                    conjunction with --prompt or --sleep to interactively
                    step through drawing commands, etc.
  --xsync, --sync   Call XSync(..., False) after each call.  Useful in
                    conjunction with --prompt or --sleep to interactively
                    step through drawing commands, etc.
  --fflush          Call fflush() after each line of output.
  --flush           Equivalent to --xflush --fflush

  --show-implementation-dependent-structs
                    Show the contents of structs that are xlib
                    implementation dependent, but still publicly
                    declared in XFree86/Xorg (may cause segfaults)


EXAMPLES
--------

Refer to the examples/ subdirectory for some sample output from running
some of these commands on my system.

* Some simple examples:

  $ xlibtrace xlogo
  $ xlibtrace xeyes
  $ xlibtrace xclock

* Tracing an xterm:

  $ xlibtrace --single xterm

* Observing how xlogo draws the logo:

  $ xlibtrace --sleep 100000 --flush xlogo
  $ xlibtrace --sleep 100000 --sync xlogo
  $ xlibtrace --prompt --flush xlogo

* Trace an xterm, which then traces an xlogo:

  $ xlibtrace xterm -e xlogo

* Trace xfig (very verbose output):

  $ xlibtrace -o xfig-xtrace.log xfig

* Trace the closed-source PDF viewer from Adobe:

  $ xlibtrace -o acroread-xtrace.log acroread


LIMITATIONS
-----------

The usual $LD_PRELOAD caveats apply, ie:

- Only works with dynamically linked binaries.

- For setuid/setgid binaries, the libxlibtrace.so shared library must be
  installed in a system location and have matching setuid/setgid permissions.


FEEDBACK
--------

Comments, feature suggestions, bug reports, patches, etc are most
welcome, please send them to Kevin Pulo <kev@pulo.com.au>.


SEE ALSO
--------

- xtrace: http://xtrace.alioth.debian.org/
- GNU xnee: http://www.gnu.org/software/xnee/

Both of these projects deal more with the X protocol, that is, the interface
between the Xlib client library and the X server.  xlibtrace deals with the
interface between the client application and the Xlib client library (ie. the
local libX11 shared library).




ChangeLog

0.2.1:
- More build system fixes.

0.2:
- Updates to allow compiling on modern systems.

0.1:
- Initial release



src/libxlibtrace-print-basic.h
/* xlibtrace: LD_PRELOAD library to trace libX11 calls
 * Copyright (C) 2007-2009 Kevin Pulo
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 * Contact:
 * Kevin Pulo
 * kev@pulo.com.au
 */


#ifndef __LIBXLIBTRACE_PRINT_BASIC_H__
#define __LIBXLIBTRACE_PRINT_BASIC_H__

#include <X11/Xlib.h>

#include "libxlibtrace-print.h"


#if defined(__LIBXLIBTRACE_PRINT_BASIC_INCLUDE_DEFINITIONS__)

/* Stuff for the .c file */
#  define __LIBXLIBTRACE_PRINT_BASIC_BODY__(x) x
#  define __LIBXLIBTRACE_PRINT_BASIC_SUFF__ 

#include <string.h>

#else

/* Stuff for the .h file */
#  define __LIBXLIBTRACE_PRINT_BASIC_BODY__(x) ;
#  define __LIBXLIBTRACE_PRINT_BASIC_SUFF__ _PROTO

#endif


#define __REALTYPE_POINTER__ void *
void print_type_POINTER_(FILE *f, const char *funcname, const char *argname, const char *type, const void * value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%p", value);
	dofflush(f);
})
#define __REALTYPE_POINTER_p__ void * *
#define __REALTYPE_POINTER_pp__ void * * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(POINTER)

#define __REALTYPE_void__ void

#define __REALTYPE_void_p__ void *
void print_type_void_p_(FILE *f, const char *funcname, const char *argname, const char *type, const void * value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%p", value);
	dofflush(f);
})
#define __REALTYPE_void_p_p__ void * *
#define __REALTYPE_void_p_pp__ void * * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(void_p)

#define __REALTYPE_int__ int
void print_type_int_(FILE *f, const char *funcname, const char *argname, const char *type, const int value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%d", value);
	dofflush(f);
})
#define __REALTYPE_int_p__ int *
#define __REALTYPE_int_pp__ int * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(int)

#define __REALTYPE_unsigned_int__ unsigned int
void print_type_unsigned_int_(FILE *f, const char *funcname, const char *argname, const char *type, const unsigned int value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%u", value);
	dofflush(f);
})
#define __REALTYPE_unsigned_int_p__ unsigned int *
#define __REALTYPE_unsigned_int_pp__ unsigned int * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(unsigned_int)

#define __REALTYPE_signed_int__ signed int
void print_type_signed_int_(FILE *f, const char *funcname, const char *argname, const char *type, const signed int value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%d", value);
	dofflush(f);
})
#define __REALTYPE_signed_int_p__ signed int *
#define __REALTYPE_signed_int_pp__ signed int * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(signed_int)

#define __REALTYPE_unsigned__ unsigned
void print_type_unsigned_(FILE *f, const char *funcname, const char *argname, const char *type, const unsigned value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%u", value);
	dofflush(f);
})
#define __REALTYPE_unsigned_p__ unsigned *
#define __REALTYPE_unsigned_pp__ unsigned * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(unsigned)

#define __REALTYPE_signed__ signed
void print_type_signed_(FILE *f, const char *funcname, const char *argname, const char *type, const signed value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%d", value);
	dofflush(f);
})
#define __REALTYPE_signed_p__ signed *
#define __REALTYPE_signed_pp__ signed * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(signed)

#define __REALTYPE_long__ long
void print_type_long_(FILE *f, const char *funcname, const char *argname, const char *type, const long value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%ld", value);
	dofflush(f);
})
#define __REALTYPE_long_p__ long *
#define __REALTYPE_long_pp__ long * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(long)

#define __REALTYPE_unsigned_long__ unsigned long
void print_type_unsigned_long_(FILE *f, const char *funcname, const char *argname, const char *type, const unsigned long value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%lu", value);
	dofflush(f);
})
#define __REALTYPE_unsigned_long_p__ unsigned long *
#define __REALTYPE_unsigned_long_pp__ unsigned long * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(unsigned_long)

#define __REALTYPE_signed_long__ signed long
void print_type_signed_long_(FILE *f, const char *funcname, const char *argname, const char *type, const signed long value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%ld", value);
	dofflush(f);
})
#define __REALTYPE_signed_long_p__ signed long *
#define __REALTYPE_signed_long_pp__ signed long * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(signed_long)

#define __REALTYPE_short__ short
void print_type_short_(FILE *f, const char *funcname, const char *argname, const char *type, const short value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%hd", value);
	dofflush(f);
})
#define __REALTYPE_short_p__ short *
#define __REALTYPE_short_pp__ short * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(short)

#define __REALTYPE_unsigned_short__ unsigned short
void print_type_unsigned_short_(FILE *f, const char *funcname, const char *argname, const char *type, const unsigned short value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%hu", value);
	dofflush(f);
})
#define __REALTYPE_unsigned_short_p__ unsigned short *
#define __REALTYPE_unsigned_short_pp__ unsigned short * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(unsigned_short)

#define __REALTYPE_signed_short__ signed short
void print_type_signed_short_(FILE *f, const char *funcname, const char *argname, const char *type, const signed short value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%hd", value);
	dofflush(f);
})
#define __REALTYPE_signed_short_p__ signed short *
#define __REALTYPE_signed_short_pp__ signed short * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(signed_short)

// Why they didn't typedef Bool as int is beyond me, ugh...
#ifdef Bool
#undef Bool
typedef int Bool;
#endif
#define __REALTYPE_Bool__ Bool
void print_type_Bool_(FILE *f, const char *funcname, const char *argname, const char *type, const Bool value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%s", (value?"True":"False"));
	dofflush(f);
})
#define __REALTYPE_Bool_p__ Bool *
#define __REALTYPE_Bool_pp__ Bool * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(Bool)

// Same for Status as for Bool
#ifdef Status
#undef Status
typedef int Status;
#endif
#define __REALTYPE_Status__ Status
void print_type_Status_(FILE *f, const char *funcname, const char *argname, const char *type, const Status value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%s", (value?"Success":"Failure"));
	dofflush(f);
})
#define __REALTYPE_Status_p__ Status *
#define __REALTYPE_Status_pp__ Status * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(Status)

#define __REALTYPE_char__ char
void print_type_char_(FILE *f, const char *funcname, const char *argname, const char *type, const char value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "'%c' (%d)", value, value);
	dofflush(f);
})

#define __REALTYPE_char_p__ char *
void print_type_char_p_(FILE *f, const char *funcname, const char *argname, const char *type, const char *value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	if (value) {
		size_t len = strlen(value);
		if (len > 30) {
			char tmp[31];
			strncpy(tmp, value, 30);
			tmp[30] = '\0';
			fprintf(f, "\"%s\"...", tmp);
		} else {
			fprintf(f, "\"%s\"", value);
		}
	} else {
		fprintf(f, "%p", value);
	}
	dofflush(f);
})
#define __REALTYPE_char_p_p__ char * *
#define __REALTYPE_char_p_pp__ char * * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(char_p)
#define __REALTYPE_char_pp__ char * *
#define __REALTYPE_char_ppp__ char * * *
// This is inappropriate, because it tries to print the string at the end of the pointer chain.
// But sometimes that causes segfaults.
//__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_PP,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(char)
// So instead just print out the pointer address.
// This doesn't work because of problems with const-ness.
//__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_ALIAS,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(char_pp, void_p)
//__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_ALIAS,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(char_ppp, void_p)
// So just do it manually for now.
void print_type_char_pp_(FILE *f, const char *funcname, const char *argname, const char *type, char ** value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%p", value);
	dofflush(f);
})
void print_type_char_ppp_(FILE *f, const char *funcname, const char *argname, const char *type, char *** value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%p", value);
	dofflush(f);
})


#define __REALTYPE_unsigned_char__ unsigned char
void print_type_unsigned_char_(FILE *f, const char *funcname, const char *argname, const char *type, const unsigned char value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "'%c' (%u)", value, value);
	dofflush(f);
})

#define __REALTYPE_unsigned_char_p__ unsigned char *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_ALIAS,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(unsigned_char_p, char_p)
#define __REALTYPE_unsigned_char_p_p__ unsigned char * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_ALIAS,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(unsigned_char_p_p, char_p_p)
#define __REALTYPE_unsigned_char_p_pp__ unsigned char * * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_ALIAS,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(unsigned_char_p_pp, char_p_pp)

#define __REALTYPE_unsigned_char_pp__ unsigned char * *
//__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_ALIAS_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(unsigned_char_p, char_p)
#define __REALTYPE_unsigned_char_ppp__ unsigned char * * *
//__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_ALIAS_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(unsigned_char_pp, char_pp)
void print_type_unsigned_char_pp_(FILE *f, const char *funcname, const char *argname, const char *type, unsigned char ** value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%p", value);
	dofflush(f);
})
void print_type_unsigned_char_ppp_(FILE *f, const char *funcname, const char *argname, const char *type, unsigned char *** value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%p", value);
	dofflush(f);
})


#define __REALTYPE_char__32___ char [32]
void print_type_char__32__(FILE *f, const char *funcname, const char *argname, const char *type, const char value[32]) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	fprintf(f, "%32s", value);
	dofflush(f);
})


#define __REALTYPE_wchar_t__ wchar_t
void print_type_wchar_t_(FILE *f, const char *funcname, const char *argname, const char *type, const wchar_t value) __LIBXLIBTRACE_PRINT_BASIC_BODY__({
	// "%lc" complains that it's a wint_t format, yet wchar_t arg, go figure
	//fprintf(f, "%li", value);
	// looks like "%lc" is fixed now...
	fprintf(f, "%lc", value);
	dofflush(f);
})

#define __REALTYPE_wchar_t_p__ wchar_t *
#define __REALTYPE_wchar_t_pp__ wchar_t * *
__INDIRECT_CALL_3__(__TRACE_PRINT_SIMPLE_TYPE_P,__LIBXLIBTRACE_PRINT_BASIC_SUFF__,__)(wchar_t)



#undef __LIBXLIBTRACE_PRINT_BASIC_BODY__
#undef __LIBXLIBTRACE_PRINT_BASIC_SUFF__


#endif



src/libxlibtrace-functions.h.sh
#!/bin/bash

# xlibtrace: LD_PRELOAD library to trace libX11 calls
# Copyright (C) 2007-2009 Kevin Pulo
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
# 
# Contact:
# Kevin Pulo
# kev@pulo.com.au


cat /usr/include/X11/Xlib.h \
| sed \
	-e 's/^#define Bool/\/\/&/' \
	-e 's/^#define Status/\/\/&/' \
	-e 's/_X_SENTINEL([0-9]*)//g' \
| gcc -C -E -P - \
| sed -e '/^[ 	]*$/d' \
| sed -e 's/,[ 	][ 	]*\.\.\.$/,\n.../' \
| awk '
{
	if (/\(\*\) \($/) {
		mode = 1;
	}
	if (/^[ 	]*\) \/\* .* \*\/,?[ 	]*$/) {
		mode = 2;
	}
	if (mode == 1) {
		$0 = gensub("\\/\\*[ 	]*[A-Za-z0-9_]*[ 	]*\\*\\/", "", "g");
		$0 = gensub("[ 	]+", " ", "g");
		printf("%s", $0);
	} else if (mode == 2) {
		printf("%s\n", gensub("[ 	]+", " ", "g"));
		mode = 0;
	} else {
		print;
	}
}
' \
| sed -e 's/[ 	]\+/ /g' \
| awk \
	-v function_parameter_overrides_file="libxlibtrace-functions-parameter-overrides" \
'
function _xlibtrace_makesafe(t) {
	#if (_xlibtrace_ispointer(t)) {
	#	return "POINTER";
	#} else {
		safet = t;
		safet = gensub("^const[ 	]+", "", 1, safet);
		safet = gensub("[ 	]+", "_", "g", safet);
		safet = gensub("[(),]", "_", "g", safet);
		safet = gensub("\\[", "_", "g", safet);
		safet = gensub("\\]", "_", "g", safet);
		safet = gensub("\\*", "p", "g", safet);
		return safet;
	#}
}

function _xlibtrace_isfunctionpointer(t) {
	return (t ~ /\(\*\)/);
}

function _xlibtrace_ispointer(t) {
	return _xlibtrace_isfunctionpointer(t) || (t ~ /\*$/);
}

function _xlibtrace_isconst(t) {
	return (t ~ /^[cC][oO][nN][sS][tT][ 	]/);
}

function _xlibtrace_isarray(t) {
	return (t ~ /\[.*\]$/);
}

function _xlibtrace_makedefn(t, n) {
	if (t == "...") {
		return sprintf("%s", t);
	} else if (_xlibtrace_isfunctionpointer(t)) {
		return gensub("^(.*\\(\\*)(\\).*)$", "\\1"n"\\2", 1, t);
	} else if (_xlibtrace_isarray(t)) {
		return gensub("^(.*)(\\[.*\\])$", "\\1"n"\\2", 1, t);
	} else {
		return sprintf("%s %s", t, n);
	}
}

function _xlibtrace_makeprotoarglist() {
	s = "(";
	for (i = 0; i < numargs; i++) {
		s = s sprintf(" %s%s", _xlibtrace_makedefn(argtype[i], argname[i]), ((i<numargs-1)?",":""));
	}
	s = s ")";
	return s;
}

function _xlibtrace_makearglist() {
	s = "(";
	for (i = 0; i < numargs; i++) {
		s = s sprintf(" %s%s", argname[i], ((i<numargs-1)?",":""));
	}
	s = s ")";
	return s;
}

function _xlibtrace_makefixedargliststart() {
	s = "(";
	for (i = 0; i < numargs - 1; i++) {
		s = s sprintf(" %s%s", argname[i], ((i<numargs-2)?",":""));
	}
	return s;
}

function _xlibtrace_finalfixedargname() {
	return argname[numargs-2];
}

BEGIN {
	numargs = 0;
	while (getline < function_parameter_overrides_file) {
		override[$1, $3] = $2;
	}
}

{
	if (/^extern [^(]*\($/) {
		funcname = gensub("^extern (.*[^A-Za-z0-9_])([A-Za-z0-9_]*)\\($", "\\2", 1);
		type = gensub("^extern (.*[^A-Za-z0-9_])([A-Za-z0-9_]*)\\($", "\\1", 1);
		type = gensub("[ 	]+$", "", "g", type);
		type = gensub("([^ *(])(\\*)", "\\1 \\2", "g", type);
	} else if ((/^)[ 	]*;$/) && funcname) {
		if (numargs == 1 && argtype[0] == "void") {
			numargs = 0;
		}
		printf("/*\n%s %s  // %d args\n", type, funcname, numargs);
		for (i = 0; i < numargs; i++) {
			printf("    \"%s\" \"%s\"\n", argtype[i], argname[i]);
		}
		printf("*/\n");

		usesvarargs = 0;
		for (i = 0; i < numargs; i++) {
			if (argtype[i] == "...") {
				usesvarargs = 1;
			}
		}

		saferettype = _xlibtrace_makesafe(type);
		if ((funcname,"") in override) {
			saferettype = override[funcname,""];
		}

		#printf("#define __TRACE_RETTYPE_%s__ %s\n", funcname, type);
		printf("#define __TRACE_SAFERETTYPE_%s__ %s\n", funcname, saferettype);
		printf("#define __TRACE_PROTOARGLIST_%s__ %s\n", funcname, _xlibtrace_makeprotoarglist());
		printf("#define __TRACE_ARGLIST_%s__ %s\n", funcname, _xlibtrace_makearglist());
		if (usesvarargs) {
			printf("#define __TRACE_FIXEDARGLISTSTART_%s__ %s\n", funcname, _xlibtrace_makefixedargliststart());
			printf("#define __TRACE_FINALFIXEDARGNAME_%s__ %s\n", funcname, _xlibtrace_finalfixedargname());
		}
		printf("#define __TRACE_ADDITIONAL_LOCAL_VARS_%s__\n", funcname);
		printf("#define __TRACE_ADDITIONAL_PRE_RUN_UNDERLYING_%s__\n", funcname);
		printf("#define __TRACE_ADDITIONAL_POST_RUN_UNDERLYING_%s__\n", funcname);

		if (numargs == 0) {
			printf("#define __TRACE_PRINTF_BODY_%s__ \n", funcname);
		} else {
			printf("#define __TRACE_PRINTF_BODY_%s__ \\\n", funcname);
			for (i = 0; i < numargs; i++) {
				if (_xlibtrace_ispointer(argtype[i]) && !_xlibtrace_isfunctionpointer(argtype[i]) && ! (argtype[i] ~ /void[ 	]+\*/) && !_xlibtrace_isconst(argtype[i])) {
					printf("    always_reprint = 1; \\\n");
					break;
				}
			}
			for (i = 0; i < numargs; i++) {
				if (argtype[i] == "...") {
					printf("    __TRACE_PRINTF_VARARG__()");
				} else {
					safeargtype = _xlibtrace_makesafe(argtype[i]);
					if ((funcname,argname[i]) in override) {
						safeargtype = override[funcname,argname[i]];
					}
					printf("    __TRACE_PRINTF_ARG__(%s, %s, %s)", funcname, safeargtype, argname[i]);
				}
				if (i<numargs-1) {
					printf(" __PRINT_COMMA__(out) \\");
				}
				printf("\n");
			}
		}

		has_display = 0;
		for (i = 0; i < numargs; i++) {
			if (argtype[i] ~ /^Display[ 	]+\*$/) {
				has_display++;
			}
		}
		if (funcname == "XFlush" || funcname == "XSync") {
			has_display = 0;
		}
		if (has_display > 0) {
			printf("#define __TRACE_RUN_UNDERLYING_EPILOGUE_%s__ \\\n", funcname);
			for (i = 0; i < numargs; i++) {
				if (argtype[i] ~ /^Display[ 	]+\*$/) {
					printf("    __TRACE_RUN_XFLUSH_XSYNC__(%s, %s, %s)", funcname, _xlibtrace_makesafe(argtype[i]), argname[i]);
					has_display--;
					if (has_display > 0) {
						printf(" \\");
					}
					printf("\n");
				}
			}
		} else {
			printf("#define __TRACE_RUN_UNDERLYING_EPILOGUE_%s__ \n", funcname);
		}

		printf("__TRACE__(%s, %s, %s)\n", ((type == "void") ? "VOID" : "TYPED"), (usesvarargs ? "VAR" : "FIXED"), funcname);
		printf("\n");
		printf("\n");

		funcname = "";
		type = "";
		numargs = 0;
	} else if (funcname) {
		argtype[numargs] = gensub("^[ 	]*(.*)[ 	]*/\\*[ 	]*([A-Za-z0-9_]*)[ 	]*\\*/,?[ 	]*$", "\\1", 1);
		argname[numargs] = gensub("^[ 	]*(.*)[ 	]*/\\*[ 	]*([A-Za-z0-9_]*)[ 	]*\\*/,?[ 	]*$", "\\2", 1);
		argtype[numargs] = gensub("^[ 	]*", "", 1, argtype[numargs]);
		argtype[numargs] = gensub("[ 	]*$", "", 1, argtype[numargs]);
		argtype[numargs] = gensub("([^ *(])(\\*)", "\\1 \\2", "g", argtype[numargs]);
		numargs = numargs + 1;
	}
}
' \
> libxlibtrace-functions.h

